File: parent/142 -- Linked List Cycle II/tutorial_2ptr.py
class Solution:
    def detectCycle(self, head: Optional[ListNode]) -> Optional[ListNode]:
        fast = slow = head
        
        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next

            if slow == fast:
                break
        else:   # only runs if while didn't break
            return None
        
        # key part. See Floyd's Cycle algo detection proof
        while head != slow:
            head, slow = head.next, slow.next
        return slow

File: parent/142 -- Linked List Cycle II/tutorial_2ptr.go
package main

/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func detectCycle(head *ListNode) *ListNode {
	slow, fast := head, head
	crossed := false

	for fast != nil && fast.Next != nil {
		slow, fast = slow.Next, fast.Next.Next

		if slow == fast {
			crossed = true
			break
		}
	}
	if !crossed {
		return nil
	}

	for head != slow {
		head, slow = head.Next, slow.Next
	}

	return slow
}
